### PR TITLE
Python: Updated Memory to retrieve documents 

### PR DESCRIPTION
**Pull request by Sarfaraz021**
# Motivation and Context
This PR introduces improvements to the memory.py file in the semantic-kernel repository. The motivation behind these changes is to enhance the functionality of the memory module by adding a method to retrieve documents. This enhancement aims to provide users with the ability to retrieve stored documents based on specific queries, thereby improving the overall usability and effectiveness of the semantic kernel.

## Description
The changes in this PR primarily focus on extending the functionality of the memory.py module to include a method for document retrieval. The approach involves implementing a new function retrieve_document() within the module, which allows users to search for and retrieve documents stored in the memory based on predefined queries.

Additionally, the PR includes modifications to the existing codebase to ensure compatibility and consistency with the newly introduced functionality. These modifications include updating the populate_memory_from_files() function to populate the memory with sample documents and enhancing the setup_chat_with_memory() function to demonstrate document retrieval in the chat interface.

Overall, these changes contribute to enriching the semantic kernel's memory capabilities, empowering users to store and retrieve information effectively during interactions with the chatbot.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -2,20 +2,13 @@
   "version": "0.2.0",
   "configurations": [
```
Comment (moonbox3) [inline .vscode/launch.json:3]: Please revert these changes.

File: .vscode/settings.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
   "[csharp]": {
     "editor.defaultFormatter": "ms-dotnettools.csharp",
     "editor.codeActionsOnSave": {
-      "source.fixAll": true
```
Comment (moonbox3) [inline .vscode/settings.json:12]: please revert all of these changes in the settings.json

File: .vscode/tasks.json [hunk 3]
```diff
@@ -1,22 +1,25 @@
 {
   "version": "2.0.0",
   "tasks": [
-    // *****************************
```
Comment (moonbox3) [inline .vscode/tasks.json:4]: please revert all of these changes in the tasks.json

File: python/notebooks/00-getting-started.ipynb [hunk 4]
```diff
@@ -49,17 +49,28 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 3,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "data": {
```
Comment (moonbox3) [inline python/notebooks/00-getting-started.ipynb:56]: please clear all outputs from Jupyter notebooks. Changes to the notebooks need to be committed without any output.

File: python/semantic_kernel/kernel.py [hunk 5]
```diff
@@ -727,9 +727,11 @@ def _create_semantic_function(
         if function_config.has_chat_prompt:
             service = self.get_ai_service(
                 ChatCompletionClientBase,
-                function_config.prompt_template_config.default_services[0]
-                if len(function_config.prompt_template_config.default_services) > 0
-                else None,
+                (
```
Comment (moonbox3) [inline python/semantic_kernel/kernel.py:730]: This is old code -- please update your local repo to latest main. This code hasn't been present since 0.5.1.dev0. We're now on 0.9.4b1.

File: python/samples/kernel-syntax-examples/memory_with_docs/data/info1.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+Your budget for 2025 is $200,000
```
Comment (moonbox3) [inline python/samples/kernel-syntax-examples/memory_with_docs/data/info1.txt:1]: Since this is a kernel example, we don't need 4 separate files that contain one line of text each. You can place the text inside of the kernel example.

File: python/samples/kernel-syntax-examples/memory_with_docs/main.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        info3_text = file.read()
+    info4_text = extract_text_from_pdf("data\info4.pdf")
+
+    await memory.save_information(collection=collection_id, id="info1", text=info1_text)
+    await memory.save_information(collection=collection_id, id="info2", text=info2_text)
+    await memory.save_information(collection=collection_id, id="info3", text=info3_text)
+    await memory.save_information(collection=collection_id, id="info4", text=info4_text)
+
+
+async def setup_chat_with_memory(
```
Comment (moonbox3) [inline python/samples/kernel-syntax-examples/memory_with_docs/main.py:45]: Our kernel examples each live in their own file. There is no need to make a sub-folder with a file `main.py`. Please follow how other kernel examples are handled.

File: python/samples/kernel-syntax-examples/memory_with_docs/main.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    )
+    kernel.add_service(embedding_gen)
+
+    memory = SemanticTextMemory(
+        storage=sk.memory.VolatileMemoryStore(), embeddings_generator=embedding_gen)
+    kernel.import_plugin_from_object(
+        TextMemoryPlugin(memory), "TextMemoryPlugin")
+
+    await populate_memory_from_files(memory)
+    # await search_memory_examples(memory)
```
Comment (moonbox3) [inline python/samples/kernel-syntax-examples/memory_with_docs/main.py:134]: Do we need this code that is commented out?

**Review by moonbox3 — CHANGES_REQUESTED** [review]
Please do not remove the contributor guidelines from the bottom of the PR description. See screenshot. 

image

**Comment by eavanvalkenburg** [comment]
@Sarfaraz021 can I also note that the PR messages does not match with the code changes. We are also looking at cleaning up and simplifying the set of samples, so I don't think this is the time to add new samples until we have a clear picture of which ones we want and need, if you really miss a sample of something that took you long to figure out, please create an issue first so that we can keep track.
